File: lib/python/opentoken/src/main/opentoken/attributes/combined_attribute.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from abc import ABC, abstractmethod
from typing import List
from opentoken.attributes.serializable_attribute import SerializableAttribute


class CombinedAttribute(SerializableAttribute, ABC):
    """
    Abstract base class for attributes that combine multiple attribute implementations.

    This class allows for combining multiple attribute implementations where each
    implementation handles a specific subset of validation and normalization logic.
    The combined attribute will iterate through all implementations until it finds
    one that reports the string as valid for validation, and uses the first
    implementation that can normalize the value for normalization.

    Subclasses must implement get_attribute_implementations() to provide
    the list of attribute implementations to combine.
    """

    @abstractmethod
    def get_attribute_implementations(self) -> List[SerializableAttribute]:
        """
        Get the list of attribute implementations to combine.

        Returns:
            List[SerializableAttribute]: The list of attribute implementations
        """
        pass

    def validate(self, value: str) -> bool:
        """
        Validate the attribute value by checking if any of the combined
        implementations consider the value valid.

        Args:
            value: The attribute value to validate

        Returns:
            True if any implementation validates the value successfully, False otherwise
        """
        if value is None:
            return False

        return any(impl.validate(value) for impl in self.get_attribute_implementations())

    def normalize(self, value: str) -> str:
        """
        Normalize the attribute value using the first implementation that
        successfully validates the value.

        Args:
            value: The attribute value to normalize

        Returns:
            The normalized value from the first validating implementation,
            or the original trimmed value if no implementation validates it
        """
        if value is None:
            return None

        for impl in self.get_attribute_implementations():
            if impl.validate(value):
                return impl.normalize(value)

        # If no implementation validates the value, return trimmed original
        return value.strip()

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/CombinedAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes;

import java.util.List;

/**
 * Abstract base class for attributes that combine multiple attribute implementations.
 * 
 * <p>
 * This class allows for combining multiple attribute implementations where each
 * implementation handles a specific subset of validation and normalization logic.
 * The combined attribute will iterate through all implementations until it finds
 * one that reports the string as valid for validation, and uses the first
 * implementation that can normalize the value for normalization.
 * </p>
 * 
 * <p>
 * Subclasses must implement {@link #getAttributeImplementations()} to provide
 * the list of attribute implementations to combine.
 * </p>
 */
public abstract class CombinedAttribute implements SerializableAttribute {

    private static final long serialVersionUID = 1L;

    /**
     * Gets the list of attribute implementations to combine.
     * 
     * @return the list of attribute implementations.
     */
    protected abstract List<SerializableAttribute> getAttributeImplementations();

    /**
     * Validates the attribute value by checking if any of the combined
     * implementations consider the value valid.
     * 
     * @param value the attribute value to validate.
     * @return true if any implementation validates the value successfully; false otherwise.
     */
    @Override
    public boolean validate(String value) {
        if (value == null) {
            return false;
        }

        return getAttributeImplementations().stream()
                .anyMatch(impl -> impl.validate(value));
    }

    /**
     * Normalizes the attribute value using the first implementation that
     * successfully validates the value.
     * 
     * @param value the attribute value to normalize.
     * @return the normalized value from the first validating implementation,
     *         or the original trimmed value if no implementation validates it.
     */
    @Override
    public String normalize(String value) {
        if (value == null) {
            return null;
        }

        for (SerializableAttribute impl : getAttributeImplementations()) {
            if (impl.validate(value)) {
                return impl.normalize(value);
            }
        }

        // If no implementation validates the value, return trimmed original
        return value.trim();
    }
}